File: csharp/CustomerBaseTest.cs
using NUnit.Framework;
using System;
using System.Collections.Generic;

namespace CustomerBase
{
	[TestFixture()]
	public class CustomerBaseTest
	{

		private Customer alice = new Customer ("Alice", "Rossi", 10000);
		private Customer bob = new Customer ("Bob", "Rossi", 20000);
		private Customer charlie = new Customer ("Charlie", "Bianchi", 30000);
		private CustomerBase customerBase;

		[SetUp]
		public void setUp ()
		{
			customerBase = new CustomerBase ();
			customerBase.Add (alice);
			customerBase.Add (bob);
			customerBase.Add (charlie);
		}

		[Test]
		public void findByLastName ()
		{
			List<Customer> found = customerBase.FindByLastName ("Rossi");
			Assert.AreEqual (new List<Customer> () { alice, bob }, found);
		}

		[Test]
		public void findByFirstAndLastName ()
		{
			List<Customer> found = customerBase.FindByFirstAndLastName ("Alice", "Rossi");
			Assert.AreEqual (new List<Customer> () { alice }, found);
		}

		[Test]
		public void findWithCreditGreaterThan ()
		{
			List<Customer> found = customerBase.FindByCreditGreaterThan (20000);
			Assert.AreEqual (new List<Customer> () { charlie }, found);
		}

		[Test]
		public void findWithCreditLessThan ()
		{
			Assert.Fail ("Open-closed violation!");
			// ...
			// We're sick and tired of adding new methods to CustomerBase.
			// CHALLENGE: can you refactor CustomerBase and its tests
			// so that new kinds of queries can be implemented WITHOUT CHANGING CUSTOMERBASE ???
		}
	}
}

File: python/customer_base_test.py
import unittest
from customer_base import CustomerBase
from customer import Customer

class CustomerBaseTest(unittest.TestCase):

    def setUp(self):
        self.alice = Customer("Alice", "Rossi", 10000)
        self.bob = Customer("Bob", "Rossi", 20000)
        self.charlie = Customer("Charlie", "Bianchi", 30000)
        self.subject = CustomerBase()
        self.subject.add(self.alice)
        self.subject.add(self.bob)
        self.subject.add(self.charlie)

    def testFindByLastName(self):
        found = self.subject.findByLastName("Rossi")
        self.assertEqual([self.alice, self.bob], found)

    def testFindByFirstAndLastName(self):
        found = self.subject.findByFirstAndLastName("Alice", "Rossi")
        self.assertEqual([self.alice], found)

    def testFindWithCreditGreaterThan(self):
        found = self.subject.findByCreditGreaterThan(20000)
        self.assertEqual([self.charlie], found)

    def testFindWithCreditLessThan(self):
        self.fail("open-closed violation")
        #
        # We're sick and tired of adding new methods to CustomerBase.
        # CHALLENGE: can you refactor CustomerBase and its tests
        # so that new kinds of queries can be implemented WITHOUT CHANGING CUSTOMERBASE ???
        #

